<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good)</span></div>

## 1 Interrupting the flow

In [1]:
for i in range(10):
    if i > 6: break
    print(i)

0
1
2
3
4
5
6


In [2]:
for i in range(10):
    if i == 6: continue
        
    print(i)

0
1
2
3
4
5
7
8
9


In [3]:
for i in range(10):
    if i % 2 == 0: continue
    
    print(i)
    
#So break will completely break the loop and move on to next command;
#Whereas, continue will simply skip that iteration if the conditions are met, and move to next iteration

1
3
5
7
9


In [14]:
number = 0

while True:
    print(number)
    number += 1
    if number > 4: break
    

#Might need someone to explain to me why I get stuck in an infinity loop when I change break to continue...
#I thought it shld be fine since it'll stop at 4

0
1
2
3
4


## 2 List comprehension!

### 2.1 Basic syntax

In [5]:
[number for number in range(10)]

[number ** 2 for number in range(10)]

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

### 2.2 List comprehension with conditions

In [6]:
[number for number in range(10) if number % 2 == 0]

[0, 2, 4, 6, 8]

## 3 Other useful stuff

### 3.1 for with unpacking

In [7]:
x, y, z = [1, 2, 3]
print(f"x = {x}, y = {y}, z = {z}")

x = 1, y = 2, z = 3


In [8]:
py_superhero_info = [["Natasha Romanoff", "Black Widow"],
                     ["Tony Stark", "Iron Man"],
                     ["Stephen Strange", "Doctor Strange"]]

for real_name, super_name in py_superhero_info:
    print(f"{real_name} is Marvel's {super_name}!")
    
#Must write the variables in the correct order of the 2D list!

Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


### 3.2 for with a zip()

In [9]:
super_names = ["Black Widow", "Iron Man", "Doctor Strange"]
real_names = ["Natasha Romanoff", "Tony Stark", "Stephen Strange"]

for real_name, super_name in zip(real_names, super_names):
    print(f"{real_name} is Marvel's {super_name}!")

Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


### 3.3 for with dictionaries

In [10]:
superhero_info = {
    "Natasha Romanoff": "Black Widow",
    "Tony Stark": "Iron Man",
    "Stephen Strange": "Doctor Strange"
}

for key, value in superhero_info.items():
    print(f"{key} is Marvel's {value}!")

#items() spits out both key and corresponding value. 


Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


In [11]:
for key in superhero_info.keys():
    value = superhero_info[key]
    print(f"{key} is Mavel's {value}!")

Natasha Romanoff is Mavel's Black Widow!
Tony Stark is Mavel's Iron Man!
Stephen Strange is Mavel's Doctor Strange!


## Exercise 1 :  Make me an odd list

In [4]:
square_odd = []
for i in range(10):
    if i % 2 == 0: continue 
    square_odd.append(i ** 2)

square_odd


[1, 9, 25, 49, 81]

## Exercise 2 :  Make me another odd list

In [1]:
[number ** 2 for number in range(10) if number % 2 != 0]

#single % is used to check one line, %% is used for the entire cell


[1, 9, 25, 49, 81]

## Exercise 3 :  Time me!

In [5]:
%%timeit 
square_odd = []
for i in range(10):
    if i % 2 == 0: continue 
    square_odd.append(i ** 2)

square_odd


1.96 µs ± 11.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [6]:
%timeit [number ** 2 for number in range(10) if number % 2 != 0]

1.93 µs ± 10.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Exercise 1 used 1.96 microsec $\pm$ 11.6ns per loop

Exercise 2 used 1.93 microsec $\pm$ 10.8ns per loop

Hence, a list comprehension is slightly faster with a higher precision too. 